pkg/resources/server.go

```diff
@@ -358,3 +366,75 @@ func (rs *resourceServer) NotifyRegistrationStatus(ctx context.Context, regstat
 	}
 	return &registerapi.RegistrationStatusResponse{}, nil
 }
+
+func (rs *resourceServer) UpdateDevices(devices []types.PciNetDevice) {
+	// Get device spec
+	deviceSpec := getDevicesSpec(devices)
+
+	// If not devices not changed skip
+	if !devicesChanged(rs.deviceSpec, deviceSpec) {
+		log.Printf("no changes to devices for \"%s\"", rs.resourceName)
+		log.Println("exposing devices: ", rs.devs)
+		return
+	}
+
```

Comment:
the operation of updating `rs.deviceSpec` and `rs.devs` is not atomic so if an `allocate()` request is received while UpdateDevices occurs, it may not behave correctly.

I would add a mutex to resourceServer to protect access for both `rs.deviceSpec` and `rs.devs` then use it where these data structures are read/written